### ISSUE TITLE
Elvish hangs up after typing 2 chars

### ISSUE THREAD

**Issue by dcu** [elves/elvish#379]
The first typed char is printed in the screen in red, after typing the second char it doesn't appear in the screen. If I press ctrl-c the first char is deleted but everything else is hang up. 
Press ctrl+\ prints the following ST:

```
goroutine 21 [running]:                                                               
github.com/elves/elvish/sys.DumpStack(0xc420166120, 0xc420029f98)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/sys/dumpstack.go:7 +0xa7
github.com/elves/elvish/shell.handleUsr1AndQuit.func1(0xc420166120)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:192 +0x86
created by github.com/elves/elvish/shell.handleUsr1AndQuit
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:190 +0xd8

goroutine 1 [runnable]:
fmt.(*pp).catchPanic(0xc42019e000, 0x4588520, 0xc4203da050, 0x73)
	/opt/go/src/fmt/print.go:523 +0x528
fmt.(*pp).handleMethods(0xc42019e000, 0xc400000073, 0x1)
	/opt/go/src/fmt/print.go:597 +0x292
fmt.(*pp).printArg(0xc42019e000, 0x4588520, 0xc4203da050, 0xc400000073)
	/opt/go/src/fmt/print.go:679 +0x171
fmt.(*pp).doPrintf(0xc42019e000, 0x4618647, 0x11, 0xc42012d908, 0x1, 0x1)
	/opt/go/src/fmt/print.go:996 +0x15a
fmt.Sprintf(0x4618647, 0x11, 0xc42012d908, 0x1, 0x1, 0x3, 0xc42012d718)
	/opt/go/src/f
```
or

```
goroutine 7 [running]:                                                                 
github.com/elves/elvish/sys.DumpStack(0xc42016a120, 0xc42002ef98)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/sys/dumpstack.go:7 +0xa7
github.com/elves/elvish/shell.handleUsr1AndQuit.func1(0xc42016a120)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:192 +0x86
created by github.com/elves/elvish/shell.handleUsr1AndQuit
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:190 +0xd8

goroutine 1 [runnable]:
github.com/elves/elvish/edit.(*Editor).addTip(0xc4200a7b00, 0x4618647, 0x11, 0xc420153908, 0x1, 0x1)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/edit/editor.go:128 +0x184
github.com/elves/elvish/edit.(*Editor).ReadLine(0xc4200a7b00, 0x0, 0x0, 0x0, 0x0)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/edit/editor.go:384 +0x73a
github.com/elves/elvish/shell.interact.func1(0x459f840, 0xc4201b4780, 0x1f, 0x4612101)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/
```
or

```
goroutine 23 [running]:                                                                 
github.com/elves/elvish/sys.DumpStack(0xc4200842a0, 0xc42002a798)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/sys/dumpstack.go:7 +0xa7
github.com/elves/elvish/shell.handleUsr1AndQuit.func1(0xc4200842a0)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:192 +0x86
created by github.com/elves/elvish/shell.handleUsr1AndQuit
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:190 +0xd8

goroutine 1 [runnable]:
syscall.CloseOnExec(0x7)
	/opt/go/src/syscall/exec_unix.go:96 +0x4c
os.Pipe(0xc42014d680, 0xc4203a40c0, 0x20, 0xc4203a6320)
	/opt/go/src/os/pipe_bsd.go:23 +0x15d
github.com/elves/elvish/eval.pcaptureOutput(0xc42014d680, 0xc420398100, 0xffffffffffffffff, 0xffffffffffffffff, 0x4588520, 0xc42014d568, 0x)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/eval/compile_value.go:349 +0x178
github.com/elves/elvish/eval.(*EvalCtx).PCaptureOutput(0xc42014d680, 0x5b00000, 0xc42000c0
```

or
```
goroutine 20 [running]:                                                                
github.com/elves/elvish/sys.DumpStack(0xc42016e120, 0xc420029798)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/sys/dumpstack.go:7 +0xa7
github.com/elves/elvish/shell.handleUsr1AndQuit.func1(0xc42016e120)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:192 +0x86
created by github.com/elves/elvish/shell.handleUsr1AndQuit
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:190 +0xd8

goroutine 1 [runnable]:
fmt.(*pp).handleMethods(0xc4203c2180, 0xc400000073, 0x4044b00)
	/opt/go/src/fmt/print.go:586 +0x103
fmt.(*pp).printArg(0xc4203c2180, 0x4588520, 0xc4203d4250, 0x73)
	/opt/go/src/fmt/print.go:679 +0x171
fmt.(*pp).doPrintf(0xc4203c2180, 0x4618647, 0x11, 0xc42014f908, 0x1, 0x1)
	/opt/go/src/fmt/print.go:996 +0x15a
fmt.Sprintf(0x4618647, 0x11, 0xc42014f908, 0x1, 0x1, 0x3, 0xc42014f718)
	/opt/go/src/fmt/print.go:196 +0x66
github.com/elves/elvish/edit.(*Editor).addTip(0xc4201ba240, 0x4618647, 0x11, 0xc4201
```
...

**Comment by dcu** [comment]
got a better ST:

```
github.com/elves/elvish/sys.DumpStack(0xc4201b8060, 0xc42002e798)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/sys/dumpstack.go:7 +0xa7
github.com/elves/elvish/shell.handleUsr1AndQuit.func1(0xc4201b8060)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:192 +0x86
created by github.com/elves/elvish/shell.handleUsr1AndQuit
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:190 +0xd8

goroutine 1 [select]:
github.com/elves/elvish/edit.(*Editor).ReadLine(0xc4200a7b00, 0x0, 0x0, 0x0, 0x0)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/edit/editor.go:364 +0x4c3
github.com/elves/elvish/shell.interact.func1(0x459f840, 0xc4201f0960, 0x1f, 0x4612101)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:133 +0x2a
github.com/elves/elvish/shell.interact(0xc420138280, 0xc420158f00)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:144 +0x148
github.com/elves/elvish/shell.(*Shell).Run(0xc420143f10, 0xc420078180, 0x0, 0x0, 0x0)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:57 +0x191
main.main()
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/main.go:143 +0x477

goroutine 4 [syscall]:
os/signal.signal_recv(0x4879a00)
	/opt/go/src/runtime/sigqueue.go:131 +0xa7
os/signal.loop()
	/opt/go/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.0
	/opt/go/src/os/signal/signal_unix.go:28 +0x41

goroutine 5 [chan receive]:
github.com/elves/elvish/eval.init.2.func1()
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/eval/port.go:56 +0x67
created by github.com/elves/elvish/eval.init.2
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/eval/port.go:55 +0x49

goroutine 6 [IO wait]:
internal/poll.runtime_pollWait(0x5960f00, 0x72, 0x0)
	/opt/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc42015a118, 0x72, 0xffffffffffffff00, 0x4876ec0, 0x48728d0)
	/opt/go/src/internal/poll/fd_poll_runtime.go:85 +0xae
internal/poll.(*pollDesc).waitRead(0xc42015a118, 0xc4201a7000, 0x1000, 0x1000)
	/opt/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc42015a100, 0xc4201a7000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/opt/go/src/internal/poll/fd_unix.go:125 +0x17b
net.(*netFD).Read(0xc42015a100, 0xc4201a7000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/opt/go/src/net/fd_unix.go:202 +0x52
net.(*conn).Read(0xc42000e038, 0xc4201a7000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/opt/go/src/net/net.go:176 +0x6d
bufio.(*Reader).Read(0xc420060120, 0xc42001a540, 0x1, 0x9, 0xc42013fca8, 0x402ba08, 0x4630f98)
	/opt/go/src/bufio/bufio.go:213 +0x30b
io.ReadAtLeast(0x4874000, 0xc420060120, 0xc42001a540, 0x1, 0x9, 0x1, 0xc420024600, 0xc42001a510, 0x182)
	/opt/go/src/io/io.go:307 +0x86
io.ReadFull(0x4874000, 0xc420060120, 0xc42001a540, 0x1, 0x9, 0xc42013fd10, 0x400681b, 0x3)
	/opt/go/src/io/io.go:325 +0x58
encoding/gob.decodeUintReader(0x4874000, 0xc420060120, 0xc42001a540, 0x9, 0x9, 0xc42013fd88, 0x402b349, 0xc400000008, 0xc42014c180)
	/opt/go/src/encoding/gob/decode.go:120 +0x63
encoding/gob.(*Decoder).recvMessage(0xc42015a300, 0xc42013fda0)
	/opt/go/src/encoding/gob/decoder.go:76 +0x57
encoding/gob.(*Decoder).decodeTypeSequence(0xc42015a300, 0x4632700, 0xc42015a300)
	/opt/go/src/encoding/gob/decoder.go:140 +0x14c
encoding/gob.(*Decoder).DecodeValue(0xc42015a300, 0x454bc00, 0xc42013a510, 0x16, 0x0, 0x0)
	/opt/go/src/encoding/gob/decoder.go:208 +0xe6
encoding/gob.(*Decoder).Decode(0xc42015a300, 0x454bc00, 0xc42013a510, 0x0, 0x0)
	/opt/go/src/encoding/gob/decoder.go:185 +0x16d
net/rpc.(*gobClientCodec).ReadResponseHeader(0xc420158ff0, 0xc42013a510, 0xc42001a510, 0x0)
	/opt/go/src/net/rpc/client.go:223 +0x45
net/rpc.(*Client).input(0xc4200601e0)
	/opt/go/src/net/rpc/client.go:109 +0xb5
created by net/rpc.NewClientWithCodec
	/opt/go/src/net/rpc/client.go:201 +0xc5

goroutine 34 [select, locked to thread]:
runtime.gopark(0x4632720, 0x0, 0x4611de7, 0x6, 0x18, 0x1)
	/opt/go/src/runtime/proc.go:277 +0x12c
runtime.selectgo(0xc42003ff50, 0xc4201b8120)
	/opt/go/src/runtime/select.go:395 +0x1138
runtime.ensureSigM.func1()
	/opt/go/src/runtime/signal_unix.go:511 +0x1fe
runtime.goexit()
	/opt/go/src/runtime/asm_amd64.s:2337 +0x1

goroutine 20 [runnable]:
github.com/elves/elvish/shell.logSignals.func1(0xc4201700c0)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:181 +0xbf
created by github.com/elves/elvish/shell.logSignals
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/shell/shell.go:180 +0x8a

goroutine 21 [IO wait]:
internal/poll.runtime_pollWait(0x5960cc0, 0x72, 0x0)
	/opt/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc420136118, 0x72, 0xffffffffffffff00, 0x4876ec0, 0x48728d0)
	/opt/go/src/internal/poll/fd_poll_runtime.go:85 +0xae
internal/poll.(*pollDesc).waitRead(0xc420136118, 0xc4201e5000, 0x1000, 0x1000)
	/opt/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc420136100, 0xc4201e5000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/opt/go/src/internal/poll/fd_unix.go:125 +0x17b
net.(*netFD).Read(0xc420136100, 0xc4201e5000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/opt/go/src/net/fd_unix.go:202 +0x52
net.(*conn).Read(0xc42008e518, 0xc4201e5000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/opt/go/src/net/net.go:176 +0x6d
bufio.(*Reader).Read(0xc4200961e0, 0xc420084650, 0x1, 0x9, 0xc4201efca8, 0x402ba08, 0x4630f98)
	/opt/go/src/bufio/bufio.go:213 +0x30b
io.ReadAtLeast(0x4874000, 0xc4200961e0, 0xc420084650, 0x1, 0x9, 0x1, 0xc420023300, 0xc420084578, 0x182)
	/opt/go/src/io/io.go:307 +0x86
io.ReadFull(0x4874000, 0xc4200961e0, 0xc420084650, 0x1, 0x9, 0xc4201efd10, 0x400681b, 0x3)
	/opt/go/src/io/io.go:325 +0x58
encoding/gob.decodeUintReader(0x4874000, 0xc4200961e0, 0xc420084650, 0x9, 0x9, 0xc4201efd88, 0x402b349, 0xc400000008, 0xc420086040)
	/opt/go/src/encoding/gob/decode.go:120 +0x63
encoding/gob.(*Decoder).recvMessage(0xc420136300, 0xc4201efda0)
	/opt/go/src/encoding/gob/decoder.go:76 +0x57
encoding/gob.(*Decoder).decodeTypeSequence(0xc420136300, 0x4632700, 0xc420136300)
	/opt/go/src/encoding/gob/decoder.go:140 +0x14c
encoding/gob.(*Decoder).DecodeValue(0xc420136300, 0x454bc00, 0xc4201593e0, 0x16, 0x0, 0x0)
	/opt/go/src/encoding/gob/decoder.go:208 +0xe6
encoding/gob.(*Decoder).Decode(0xc420136300, 0x454bc00, 0xc4201593e0, 0x0, 0x0)
	/opt/go/src/encoding/gob/decoder.go:185 +0x16d
net/rpc.(*gobClientCodec).ReadResponseHeader(0xc42007dbc0, 0xc4201593e0, 0xc420084578, 0x0)
	/opt/go/src/net/rpc/client.go:223 +0x45
net/rpc.(*Client).input(0xc4200962a0)
	/opt/go/src/net/rpc/client.go:109 +0xb5
created by net/rpc.NewClientWithCodec
	/opt/go/src/net/rpc/client.go:201 +0xc5

goroutine 23 [select]:
github.com/elves/elvish/edit/tty.(*Reader).Run(0xc42007da40)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/edit/tty/reader.go:97 +0x141
created by github.com/elves/elvish/edit.(*Editor).ReadLine
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/edit/editor.go:345 +0x1a5

goroutine 35 [select]:
github.com/elves/elvish/edit/tty.(*AsyncReader).Run(0xc420087500)
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/edit/tty/async_reader.go:107 +0x373
created by github.com/elves/elvish/edit/tty.(*Reader).Run
	/home/user1/Projects/GoCode/src/github.com/elves/elvish/edit/tty/reader.go:94 +0xa4
exit status 3
```

**Comment by xiaq** [comment]
Hi, thanks for helping debug this! The hang is caused by error from tty.asyncReader not correctly bubbled up, so tty.asyncReader is blocked when trying to send the error; I just pushed ae33603c73d58c55409bbb5696d5ad4d45780190 to fix the blocking.

Now Elvish will show you an error instead of being blocked. The fact that there is an error is still a problem :-/

I see that you found "read /dev/stdin: waiting for unsupported file type". What OS and terminal emulator are you using? Does the problem still occur if you change a terminal emulator? Also, can you update to HEAD and verify that it is consistently this error?

**Comment by dcu** [comment]
OSX sierra, iTerm. The same happens in "Terminal"

your fix didn't work. If I add a gorouting that consumes the errors from the channel it works, something like:
```go
go func() { for range <- chErr {} }()
```
which makes me think that it's producing more errors than what it is trying to read

**Comment by xiaq** [comment]
Confirmed over Telegram that @dcu was using Go 1.9 to compile Elvish and switching to Go 1.8 fixes the problem; Go 1.9 is not yet released so it is probably a bug of the compiler. I will try reproducing with Go 1.9 and report if this is indeed a compiler bug.

The blocking has been fixed by the latest few commits.

**Comment by xiaq** [comment]
The issue with go1.9 is now tracked in #384.
